xds/internal/server/listener_wrapper.go

```diff
@@ -156,24 +135,129 @@ type listenerWrapper struct {
 
 	// rdsHandler is used for any dynamic RDS resources specified in a LDS
 	// update.
 	rdsHandler *rdsHandler
-	// rdsUpdates are the RDS resources received from the management
-	// server, keyed on the RouteName of the RDS resource.
-	rdsUpdates unsafe.Pointer // map[string]xdsclient.RouteConfigUpdate
-	// rdsUpdateCh is a channel for XDSClient RDS updates.
-	rdsUpdateCh chan rdsHandlerUpdate
+}
+
+func (l *listenerWrapper) handleLDSUpdate(update xdsresource.ListenerUpdate) {
+	ilc := update.InboundListenerCfg
+	// Make sure that the socket address on the received Listener resource
+	// matches the address of the net.Listener passed to us by the user. This
+	// check is done here instead of at the XDSClient layer because of the
+	// following couple of reasons:
+	// - XDSClient cannot know the listening address of every listener in the
+	//   system, and hence cannot perform this check.
+	// - this is a very context-dependent check and only the server has the
+	//   appropriate context to perform this check.
+	//
+	// What this means is that the XDSClient has ACKed a resource which can push
+	// the server into a "not serving" mode. This is not ideal, but this is
+	// what we have decided to do.
+	if ilc.Address != l.addr || ilc.Port != l.port {
+		l.mu.Lock()
+		l.switchModeLocked(connectivity.ServingModeNotServing, fmt.Errorf("address (%s:%s) in Listener update does not match listening address: (%s:%s)", ilc.Address, ilc.Port, l.addr, l.port))
+		l.mu.Unlock()
+		return
+	}
+
+	l.pendingFilterChainManager = ilc.FilterChains
+	l.rdsHandler.updateRouteNamesToWatch(ilc.FilterChains.RouteConfigNames)
+
+	if l.rdsHandler.determineRouteConfigurationReady() {
+		l.maybeUpdateFilterChains()
+	}
+}
+
+// maybeUpdateFilterChains swaps in the pending filter chain manager to the
+// active one if the pending filter chain manager is present. If a swap occurs,
+// it also  drains (gracefully stops) any connections that were accepted on the
+// old active filter chain manager, and puts this listener in state SERVING.
+// Must be called within an xDS Client Callback.
+func (l *listenerWrapper) maybeUpdateFilterChains() {
+	if l.pendingFilterChainManager == nil {
+		// Nothing to update, return early.
+		return
+	}
+
+	l.mu.Lock()
+	l.switchModeLocked(connectivity.ServingModeServing, nil)
+	// "Updates to a Listener cause all older connections on that Listener to be
+	// gracefully shut down with a grace period of 10 minutes for long-lived
+	// RPC's, such that clients will reconnect and have the updated
+	// configuration apply." - A36
+	var connsToClose []net.Conn
+	if l.activeFilterChainManager != nil { // If there is a filter chain manager to clean up.
+		connsToClose = l.activeFilterChainManager.Conns()
+	}
+	l.activeFilterChainManager = l.pendingFilterChainManager
+	l.pendingFilterChainManager = nil
+	l.instantiateFilterChainRoutingConfigurationsLocked()
+	l.mu.Unlock()
+	for _, conn := range connsToClose {
+		conn.(*connWrapper).drain()
+	}
```

Comment:
Hmmm yeah we did discuss this with Eric since Java does this async. I don't think this is that important (doesn't seem like the Drain operation takes too long - and doesn't block anything/callback into this listener wrapper so no potential for deadlocks) but also seems like no harm in doing so since will unblock the listener wrapper operation that is being processed, allowing any other sync operations to get to processing faster. I also made the Drain call on a mode switch to serving mode not serving happen async.